titus-ext/zookeeper/src/main/java/com/netflix/titus/ext/zookeeper/supervisor/ZookeeperLeaderElector.java

```diff
@@ -143,16 +153,23 @@ private LeaderElectionProcess() {
         }
 
         private boolean leaveIfNotLeader() {
-            if (!leaderFlag) {
+            if (leaderFlag) {
                 return false;
             }
 
             // Not a leader yet, so it is safe to leave the leader election process.
             close();
+
+            if (leaderActivator.isLeader()) {
```

Comment:
isn't this still racy? The leader activation can still happen right after this check, and we'd still return true. Shouldn't this be fully serialized with `announceLeader()` instead?

I see that you added two checks on `terminateIfClosed()` to `announceLeader()`, but wouldn't it be better to just prevent activation completely?

Reply:
Good point. I will add `synchronized` statement within the leader election callback.